src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -292,6 +260,37 @@ static bool HasStateFlag(Http2OutputProducer.State state, Http2OutputProducer.St
         _log.Http2ConnectionQueueProcessingCompleted(_connectionId);
     }
 
+    private bool TryQueueProducerForConnectionWindowUpdate(long actual, Http2OutputProducer producer)
+    {
+        lock (_windowUpdateLock)
+        {
+            // Check the connection window under a lock so that we don't miss window updates
+            if (_connectionWindow == 0)
```

Comment:
I think only the stream window can go negative. That's because `SETTINGS_INITIAL_WINDOW_SIZE` can be decreased after the connection starts, but that only affects stream-level flow control. The connection window can only be updated with WINDOW_UPDATES which cannot be negative.

Reply:
Right!